govc ls:  strconv.ParseUint: parsing "-128": invalid syntax

run command govc ls xxxx(xxx is a vm url) and print  `strconv.ParseUint: parsing "-128": invalid syntax` to console.

I got the same error message in finder.ResourcePool() function.

```
☁  sandbox  go run main.go                                                                                                                                        [~/tmp/sandbox]
2015/04/10 09:34:15 [INFO] VMWare vSphere Client configured for URL: &{0xc208134840 0xc208161330}
2015/04/10 09:34:16 [DEBUG] &object.Datacenter{Common:object.Common{c:(*vim25.Client)(0xc208134840), r:types.ManagedObjectReference{Type:"Datacenter", Value:"datacenter-2"}}}
2015/04/10 09:34:16 [DEBUG] &find.Finder{client:(*vim25.Client)(0xc208134840), recurser:list.Recurser{Collector:(*property.Collector)(0xc20803bd70), All:true, TraverseLeafs:false}, dc:(*object.Datacenter)(nil), folders:(*object.DatacenterFolders)(nil)}
2015/04/10 09:34:17 [DEBUG] &object.DatacenterFolders{VmFolder:(*object.Folder)(0xc20803b650), HostFolder:(*object.Folder)(0xc20803b6e0), DatastoreFolder:(*object.Folder)(0xc20803b740), NetworkFolder:(*object.Folder)(0xc20803b770)}
2015/04/10 09:34:27 [ERROR] Error setting up client: strconv.ParseUint: parsing "-128": invalid syntax
2015/04/10 09:34:27 [DEBUG] (*object.ResourcePool)(nil)
☁  sandbox                                                                                                                                                        [~/tmp/sandbox]
```

main.go file: https://gist.github.com/tkak/4622e63aa175a4bd95fc

--- a/vim25/xml/read.go
+++ b/vim25/xml/read.go
@@ -631,9 +631,30 @@ func copyValue(dst reflect.Value, src []byte) (err error) {
 		}
 		dst.SetInt(itmp)
 	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
-		utmp, err := strconv.ParseUint(string(src), 10, dst.Type().Bits())
-		if err != nil {
-			return err
+		var utmp uint64
+		if len(src) > 0 && src[0] == '-' {
+			// Negative value for unsigned field.
+			// Assume it was serialized following two's complement.
+			itmp, err := strconv.ParseInt(string(src), 10, dst.Type().Bits())
+			if err != nil {
+				return err
+			}
+			// Reinterpret value based on type width.
+			switch dst.Type().Bits() {
+			case 8:
+				utmp = uint64(uint8(itmp))
+			case 16:
+				utmp = uint64(uint16(itmp))
+			case 32:
+				utmp = uint64(uint32(itmp))
+			case 64:
+				utmp = uint64(uint64(itmp))
+			}
+		} else {
+			utmp, err = strconv.ParseUint(string(src), 10, dst.Type().Bits())
+			if err != nil {
+				return err
+			}
 		}
 		dst.SetUint(utmp)
 	case reflect.Float32, reflect.Float64:

--- a/vim25/xml/read_test.go
+++ b/vim25/xml/read_test.go
@@ -712,3 +712,51 @@ func TestUnmarshalIntoInterface(t *testing.T) {
 		t.Errorf("failed to unmarshal into interface, have %q want %q", have, want)
 	}
 }
+
+// https://github.com/vmware/govmomi/issues/246
+func TestNegativeValuesUnsignedFields(t *testing.T) {
+	type T struct {
+		I   string
+		O   interface{}
+		U8  uint8  `xml:"u8"`
+		U16 uint16 `xml:"u16"`
+		U32 uint32 `xml:"u32"`
+		U64 uint64 `xml:"u64"`
+	}
+
+	var tests = []T{
+		{I: "<T><u8>-128</u8></T>", O: uint8(0x80)},
+		{I: "<T><u8>-1</u8></T>", O: uint8(0xff)},
+		{I: "<T><u16>-32768</u16></T>", O: uint16(0x8000)},
+		{I: "<T><u16>-1</u16></T>", O: uint16(0xffff)},
+		{I: "<T><u32>-2147483648</u32></T>", O: uint32(0x80000000)},
+		{I: "<T><u32>-1</u32></T>", O: uint32(0xffffffff)},
+		{I: "<T><u64>-9223372036854775808</u64></T>", O: uint64(0x8000000000000000)},
+		{I: "<T><u64>-1</u64></T>", O: uint64(0xffffffffffffffff)},
+	}
+
+	for _, test := range tests {
+		err := Unmarshal([]byte(test.I), &test)
+		if err != nil {
+			t.Errorf("Unmarshal error: %v", err)
+			continue
+		}
+
+		var expected = test.O
+		var actual interface{}
+		switch reflect.ValueOf(test.O).Type().Kind() {
+		case reflect.Uint8:
+			actual = test.U8
+		case reflect.Uint16:
+			actual = test.U16
+		case reflect.Uint32:
+			actual = test.U32
+		case reflect.Uint64:
+			actual = test.U64
+		}
+
+		if !reflect.DeepEqual(actual, expected) {
+			t.Errorf("Actual: %v, expected: %v", actual, expected)
+		}
+	}
+}

